### PR TITLE
fix(codebuild): Secret env variable from another account fails on Key decryption

### LINKED ISSUE #14043: (CodeBuild): Add KMS decrypt to policy for secrets from another account
#13706 added support for reading secrets from another account through specifying the secrets ARN. The KMS key would also live in the different account and it would be awesome if we could add the `kms:Decrypt` action to the policy. I am not sure if it would be necessary to check which kms key the secret uses or if the wildcard would suffice here. Meaning would this ressource suffice `arn:<SecretPartition>:kms:<SecretRegion>:<SecretAccount>:key/*`?

### Use Case

When using a secret from another account one get's the error message that `Access to KMS is not allowed`.
Makes secrets from another account available within CodeBuild without having to modify the policy.

### Proposed Solution

When granting the principal for `secretsmanager:GetSecretValue` we should also add the `kms:Decrypt` principal.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by Kruspe**
When providing a secretArn for the EnvironmentVariables allow kms:Decrypt
action for any key so that CodeBuild is able to get secrets that are stored
in different accounts.

fixes #14043

----

@skinny85 I didn't implement the if statement we talked about to check weither the secret actually lives in another account, since I wasn't able to come up with a good test for that. Any suggestions on this?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by skinny85** [comment]
Thanks for the contribution @Kruspe!

> @skinny85 I didn't implement the if statement we talked about to check weither the secret actually lives in another account, since I wasn't able to come up with a good test for that. Any suggestions on this?

It's actually very easy to write tests in the CDK for cross-account usecases - they look exactly like other unit tests, you just provide different account numbers when creating the Stacks. Here's a simple example from the CodePipeline tests: https://github.com/aws/aws-cdk/blob/8a949dc24b13f8b7da17c102501050bac7323bf7/packages/%40aws-cdk/aws-codepipeline-actions/test/pipeline.test.ts#L966-L1049

**Comment by Kruspe** [comment]
@skinny85 thanks for the hint! Should be ready for review now.

File: packages/@aws-cdk/aws-codebuild/test/test.project.ts [hunk 1]
```diff
@@ -1021,7 +1021,10 @@ export = {
 
       'can be provided as a verbatim full secret ARN followed by a JSON key'(test: Test) {
         // GIVEN
-        const stack = new cdk.Stack();
+        const app = new cdk.App();
+        const stack = new cdk.Stack(app, 'ProjectStack', {
+          env: { account: '123456789012' },
+        });
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/test/test.project.ts:1027]: Can we leave the existing tests as they are?

I'm fine with adding an assertion to them, but _not_ with changing any pre-conditions they have.

Same comment below.

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 2]
```diff
@@ -800,6 +801,20 @@ export class Project extends ProjectBase {
               account: parsedArn?.account,
               region: parsedArn?.region,
             }));
+            // if secret comes from another account, SecretsManager will need to access
+            // KMS on the other account as well to be able to get the secret
+            if (parsedArn?.account !== stack.account) {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts:806]: I'm 99% sure this condition is incorrect 🙂. If `parsedArn` is `undefined`, we should never go into this `if` (because in that case we can't know whether the Secret is from a different account), but we do today (because `stack.account` is never `undefined`).

There's also a subtlety with how we need to compare the accounts, because they can be unspecified, so roughly equivalent to the `AWS::Region` intrinsic.

In summary, this line needs to become:

```suggestion
            if (parsedArn && Token.compareStrings(parsedArn.account, stack.account) === TokenComparison.DIFFERENT) {
```

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 3]
```diff
@@ -800,6 +801,20 @@ export class Project extends ProjectBase {
...
+              kmsIamResources.push(stack.formatArn({
+                service: 'kms',
+                resource: 'key',
+                resourceName: '*',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts:810]: A comment here saying something like "We don't know the ID of the Key, but since this is a cross-account access, the Key will have to define the correct trust policy to allow this access anyway, so a `"*"` here is safe" would be nice 🙂.

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 4]
```diff
@@ -800,6 +801,20 @@ export class Project extends ProjectBase {
...
+                sep: '/',
+                // if we were given an ARN, we need to use the provided partition/account/region
+                partition: parsedArn?.partition,
+                account: parsedArn?.account,
+                region: parsedArn?.region,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts:815]: If you take my suggestion from above, this will become

```suggestion
                partition: parsedArn.partition,
                account: parsedArn.account,
                region: parsedArn.region,
```

Another good sign this is a needed change 🙂.

File: packages/@aws-cdk/aws-codebuild/lib/project.ts [hunk 5]
```diff
@@ -721,6 +721,7 @@ export class Project extends ProjectBase {
     const ret = new Array<CfnProject.EnvironmentVariableProperty>();
     const ssmIamResources = new Array<string>();
     const secretsManagerIamResources = new Array<string>();
+    const kmsIamResources = new Array<string>();
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/lib/project.ts:724]: Should this be a `Set` instead? Feels like if we provide 2 Secrets from another account, we should get only a single permissions statement in our IAM Role...

File: packages/@aws-cdk/aws-codebuild/test/test.project.ts [hunk 6]
```diff
@@ -1098,6 +1115,72 @@ export = {
...
+              'Effect': 'Allow',
+              'Resource': 'arn:aws:kms:us-west-2:123456789012:key/*',
+            }),
+          },
+        })));
+
+        test.done();
+      },
+
+      'can be provided as a verbatim partial secret ARN from another account'(test: Test) {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/test/test.project.ts:1132]:
```suggestion
      "when provided as a verbatim partial secret ARN from another account, adds permissions to decrypt Keys in the Secret's account"(test: Test) {

File: packages/@aws-cdk/aws-codebuild/test/test.project.ts [hunk 7]
```diff
@@ -1098,6 +1115,72 @@ export = {
...
+        // THEN
+        expect(stack).to(haveResourceLike('AWS::IAM::Policy', {
+          'PolicyDocument': {
+            'Statement': arrayWith({
+              'Action': 'secretsmanager:GetSecretValue',
+              'Effect': 'Allow',
+              'Resource': 'arn:aws:secretsmanager:us-west-2:901234567890:secret:mysecret*',
+            }),
+          },
+        }));
```
Comment (skinny85) [inline packages/@aws-cdk/aws-codebuild/test/test.project.ts:1171]: Let's remove those two assertions, I think we check for this plenty already in the other tests.

**Review by skinny85 — DISMISSED** [review]
Looks great @Kruspe! Couple of minor changes before we merge this in. Thanks for the contribution!

File: packages/@aws-cdk/aws-codebuild/lib/project.ts (hunk 5 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts:724]: You know, there's a chance the IAM library will do this for us.

A unit test will tell you if I'm right, hint hint 😉

File: packages/@aws-cdk/aws-codebuild/test/test.project.ts (hunk 1 above)
Comment (Kruspe) [reply packages/@aws-cdk/aws-codebuild/test/test.project.ts]: Yeah we won't need this with the change to suggested below 👍🏿

File: packages/@aws-cdk/aws-codebuild/lib/project.ts (hunk 5 above)
Comment (Kruspe) [reply packages/@aws-cdk/aws-codebuild/lib/project.ts]: Wrote a test and then needed a set 😄

**Review by skinny85 — APPROVED** [review]
Looks awesome @Kruspe, thanks so much for the contribution!
